log: clean up the handling of logging in tests.

knz opened the pull request:
Prior to this patch, if no logging output directory is specified and
there is no store configured on the command line (the default in most
tests), logging would silently create an output directory in the
system-wide temporary directory. This caused test logging output to be
undiscoverable, and, since this directory was never removed, would
cause stress test to exhaust the filesystem's inode supply.

This patch improves upon this behavior as follows:

- a logging output directory is not any more created automatically.

- during normal logging, if no output directory is set, logging goes
  to stderr automatically instead.

- a new struct TestLogScope is introduced, intended to be used by
  tests to scope their logging output to the duration of the test.
  This automatically creates a temporary directory named after the
  test and removes it when the test complete, only when the test has
  failed.

- the tests in `pkg/util/log` and `pkg/cli` that need it are modified
  to use `TestLogScope`.

- as a side effect, the tests in pkg/cli are simplified, and a bug in
  the implementation of TestUserInsecure that would
  non-deterministically cause TestDumpRow to fail has been fixed.

Fixes #10670.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10675)
<!-- Reviewable:end -->

knz committed:
log: cleanup the handling of logging in tests.

Prior to this patch, if no logging output directory is specified and
there is no store configured on the command line (the default in most
tests), logging would silently create an output directory in the
system-wide temporary directory. This caused test logging output to be
undiscoverable, and, since this directory was never removed, would
cause stress test to exhaust the filesystem's inode supply.

This patch improves upon this behavior as follows:

- a logging output directory is not any more created automatically.

- during normal logging, if no output directory is set, logging goes
  to stderr automatically instead.

- a new struct testLogScope is introduced, intended to be used by
  util/log tests to scope their logging output to the duration of the
  test. This automatically creates a temporary directory named after
  the test and removes it when the test complete, only when the test
  has failed.

- as a side effect, the tests in pkg/cli are simplified, and a bug in
  the implementation of TestUserInsecure that would
  non-deterministically cause TestDumpRow to fail has been fixed.

petermattis commented:
> Prior to this patch, logging would create a directory in the system-wide temporary directory for logging if no directory was otherwise set. 

When was this being used? `cockroach start` outputs to `cockroach-data/logs` by default.

knz commented:
It was used in tests only, as far as I could see.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comment.

---

Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYACkoRePbJAXdQZmq:-KWYACkoRePbJAXdQZmr:bpgsd53) ([raw file](https://github.com/cockroachdb/cockroach/blob/43c00912da9811412ce33d7fd37ec8913b00c5e0/pkg/cli/cli_test.go#L71)):*
> ```Go
> 	cliCtx.InitCLIDefaults()
> 
> 	log.DisableLogFileOutput()
> ```

This disables log file output for this test and all subsequent tests. That doesn't seem good. I'm also not clear on why this was necessary.

---

*[pkg/cli/cli_test.go, line 257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWY9p1b0od3UOdouOPn:-KWY9p1b0od3UOdouOPo:burwlug) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b5c4fd92d9479ed14c55890f680b8986427c676/pkg/cli/cli_test.go#L257)):*
> ```Go
> 	// preserving the logs on failure).
> 	c.cleanupFunc()
> 	security.SetReadFileFn(securitytest.Asset)
> ```

This is different than what the test was previously doing. Now you're always running the cleanup functions (even on failure), but not running the stopper. I'm not clear why this portion of the change was necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWYA-Hrm63B61uNoYhB:b-913mt2)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r1, 6 of 8 files at r2.
Review status: 7 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/util/log/file.go, line 62 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYS_00LLAuQeCFo-vi:-KWYS_00LLAuQeCFo-vj:bd88xal) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc79ff1f81baedb41cc6528e824ae9b688f9c6c/pkg/util/log/file.go#L62)):*
> ```Go
> 	defer l.Unlock()
> 	if l.name != "" {
> 		return fmt.Errorf("log directory already set to %s", l.name)
> ```

errors.Errorf throughout

---

*[pkg/util/log/file.go, line 93 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYSRTt1ZBts3pkDR99:-KWYSRTt1ZBts3pkDR9A:b-6i9zpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc79ff1f81baedb41cc6528e824ae9b688f9c6c/pkg/util/log/file.go#L93)):*
> ```Go
> }
> 
> func (l *logDirName) get() string {
> ```

why is there both this function and `isSet` above instead of `get` which returns an error?

---

*[pkg/util/log/flags.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYRniW8JDaXlDU1jfX:-KWYRniW8JDaXlDU1jfY:b-2ovx3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b5c4fd92d9479ed14c55890f680b8986427c676/pkg/util/log/flags.go#L27)):*
> ```Go
> func init() {
> 	logflags.InitFlags(&logging.mu, &logging.toStderr,
> 		newStringValue(&logDir, &logDirSet), &logging.nocolor, &logging.verbosity,
> ```

logDirSet is no longer used, it seems.

---

*[pkg/util/log/log.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYSHavN9GWB3gNqCRu:-KWYSHavN9GWB3gNqCRv:bvtrhcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc79ff1f81baedb41cc6528e824ae9b688f9c6c/pkg/util/log/log.go#L56)):*
> ```Go
> // For unittesting only.
> func EnableLogFileOutput(dir string) error {
> 	if err := logDir.Set(dir); err != nil {
> ```

this can just move to the bottom of the function and be `return loDir.Set(dir)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 5 of 12 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/cli/cli_test.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYACkoRePbJAXdQZmq:-KWYfCVz_6hdEQi1Lw2U:b-btuahq) ([raw file](https://github.com/cockroachdb/cockroach/blob/43c00912da9811412ce33d7fd37ec8913b00c5e0/pkg/cli/cli_test.go#L71)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This disables log file output for this test and all subsequent tests. That doesn't seem good. I'm also not clear on why this was necessary.
</details>
Yes this was an oversight. The proper fix requires to equip all tests with the necessary code to set up the logging directory. Done.

---

*[pkg/cli/cli_test.go, line 257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWY9p1b0od3UOdouOPn:-KWYfJ7YV-pB-1YPKWRu:b-f1c3hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b5c4fd92d9479ed14c55890f680b8986427c676/pkg/cli/cli_test.go#L257)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is different than what the test was previously doing. Now you're always running the cleanup functions (even on failure), but not running the stopper. I'm not clear why this portion of the change was necessary.
</details>
Agreed; fixed.

---

*[pkg/util/log/file.go, line 62 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYS_00LLAuQeCFo-vi:-KWYfKkl69GgRasqUYJC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc79ff1f81baedb41cc6528e824ae9b688f9c6c/pkg/util/log/file.go#L62)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> errors.Errorf throughout
</details>
Done.

---

*[pkg/util/log/file.go, line 93 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYSRTt1ZBts3pkDR99:-KWYgFs15GhBBySPeNJ_:bz5py0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc79ff1f81baedb41cc6528e824ae9b688f9c6c/pkg/util/log/file.go#L93)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is there both this function and `isSet` above instead of `get` which returns an error?
</details>
Good idea! Done.

---

*[pkg/util/log/flags.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYRniW8JDaXlDU1jfX:-KWYgHd07jJ5ssLc48Ko:b-dwsdv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b5c4fd92d9479ed14c55890f680b8986427c676/pkg/util/log/flags.go#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> logDirSet is no longer used, it seems.
</details>
That's correct. Why do you highlight so?

---

*[pkg/util/log/log.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYSHavN9GWB3gNqCRu:-KWYgf6YZAmibu-ZbSXk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc79ff1f81baedb41cc6528e824ae9b688f9c6c/pkg/util/log/log.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this can just move to the bottom of the function and be `return loDir.Set(dir)`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
@tamird can you help with this, there are two issues that remain which I'm not sure how to best address:

- the dump test in pkg/cli now fails and I'm not sure why
- TestDockerReadWriteBidirectionalReferenceVersion and its Forward friend need to be plumbed with a log scope, but I'm not sure where to best instrument this.

tamird commented:
Reviewed 2 of 8 files at r3.
Review status: 6 of 12 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/flags.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYRniW8JDaXlDU1jfX:-KWYjoY6TdmwfY0qO8s_:bg12xzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b5c4fd92d9479ed14c55890f680b8986427c676/pkg/util/log/flags.go#L27)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> That's correct. Why do you highlight so?
</details>
Because it still seems to exist.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 12 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 209 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZAwp0FTN7vYc_6ucq:-KWZAwp0FTN7vYc_6ucr:b8y0a80) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/cli/cli_test.go#L209)):*
> ```Go
> func TestQuit(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	l := log.MakeTestLogScope(t.Fatal)
> ```

Do all of the cli tests need a log scope? Could you put the log scope stuff inside of `newCLITest`?

---

*[pkg/util/log/file.go, line 210 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZCMvloLOCQW8qpELE:-KWZCMvloLOCQW8qpELF:b-tpzg63) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/util/log/file.go#L210)):*
> ```Go
> // errors.
> func create(severity Severity, t time.Time) (f *os.File, filename string, err error) {
> 	dir, err := logDir.get()
> ```

Rather than erroring if we can't log to a file, can we change the higher-level logging code to do `--logtostderr` if no log directory is specified. I think this would remove the need for a bunch (but not all) of the `TestLogScope` stuff. And it makes more sense to me: if you don't specify a log directory we'll output the logs somewhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

tamird commented:
Can you expand the commit message? There's some complex infrastructure you're adding here, and I'd like to understand why it's so complex (rather than simply `defer doStuffAndReturnClosure()()` in relevant tests).

---

Review status: 6 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/test_log_scope.go, line 39 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZD15gUn-3Zgu4fvWt:-KWZD15gUn-3Zgu4fvWu:b-cgfhak) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/util/log/test_log_scope.go#L39)):*
> ```Go
> // package testing.
> func MakeTestLogScope(fatalFn func(...interface{})) TestLogScope {
> 	pc, _, _, ok := runtime.Caller(1)
> ```

util/caller.Lookup ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWZE5YG3uiTiKvNl2yi:bw2q435)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYRs! 

I have extended the commit message and added several comments in the code. Let me know what you think.

---

Review status: 6 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/cli/cli_test.go, line 209 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZAwp0FTN7vYc_6ucq:-KWc-CfVTbClDdKEDweJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/cli/cli_test.go#L209)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Do all of the cli tests need a log scope? Could you put the log scope stuff inside of `newCLITest`?
</details>
Done.

---

*[pkg/util/log/file.go, line 210 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZCMvloLOCQW8qpELE:-KWcSse3wu8dxaTUnJr3:b1shk9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/util/log/file.go#L210)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Rather than erroring if we can't log to a file, can we change the higher-level logging code to do `--logtostderr` if no log directory is specified. I think this would remove the need for a bunch (but not all) of the `TestLogScope` stuff. And it makes more sense to me: if you don't specify a log directory we'll output the logs somewhere.
</details>
I'm not sure exactly how we would want this to work. One issue is that many tests can run one after another in the same process; and the logging settings for one test may not be appropriate for the next. How to detect the difference between "the log directory has been specified" and "the previous test has sent its logging to some directory"? Or between "the log directory has not been specified" and "the previous test has disabled logging to file to see if it was sent to stderr properly"? 

---

*[pkg/util/log/flags.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWYRniW8JDaXlDU1jfX:-KWc43kUG-bP0pE4SwDB:b-v30qn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b5c4fd92d9479ed14c55890f680b8986427c676/pkg/util/log/flags.go#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Because it still seems to exist.
</details>
`git grep logDirSet` is empty for me.

There's a separate method `DirSet()` which is used in `start.go` to override the log directory to the first store directory if no log dir was set by a flag. I think we still want that.

---

*[pkg/util/log/test_log_scope.go, line 39 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZD15gUn-3Zgu4fvWt:-KWcSpUSMdBD-p6rJIWY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/util/log/test_log_scope.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> util/caller.Lookup ?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWcTRsT9H-sQT94Yn3-:bgflqtu)*
<!-- Sent from Reviewable.io -->

tamird commented:
You changed a bunch of Example_foo to TestFoo - this changes the semantics. It is now the case that the comment at the end of the test is just dead, it is no longer compared to the example's stdout. See https://golang.org/pkg/testing/#pkg-examples.

---

Review status: 6 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWcUy5yOYQs3twmHTVf:b912hd7)*
<!-- Sent from Reviewable.io -->

knz commented:
Oh that's what the examples were doing!
I didn't know this.

Now then how do you suggest to report an error from inside an Example function?
Like what should I call Fatal() on?

tamird commented:
log.Fatal.

On Tue, Nov 15, 2016 at 10:02 AM, kena <notifications@github.com> wrote:

> Oh that's what the examples were doing!
> I didn't know this.
>
> Now then how do you suggest to report an error from inside an Example
> function?
> Like what should I call Fatal() on?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10675#issuecomment-260664580>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPETZNHBrhfbl3CO8IErzUgMfyotWks5q-cl4gaJpZM4KxV_w>
> .
>

petermattis commented:
Review status: 6 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/file.go, line 210 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZCMvloLOCQW8qpELE:-KWcXEwUaeOsodG6i7L3:b4l10w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/util/log/file.go#L210)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'm not sure exactly how we would want this to work. One issue is that many tests can run one after another in the same process; and the logging settings for one test may not be appropriate for the next. How to detect the difference between "the log directory has been specified" and "the previous test has sent its logging to some directory"? Or between "the log directory has not been specified" and "the previous test has disabled logging to file to see if it was sent to stderr properly"? 
</details>
In `loggingT.outputLogEntry`, consider `!logDir.isSet()` to be equivalent to `l.toStderr`. Tests usually log to stderr by default anyone. The ones that don't should be special snowflakes which use `TestLogScope`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 6 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/file.go, line 210 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWZCMvloLOCQW8qpELE:-KWciXxavk1R30psxdA9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac3ab815d11141dadee0b4d44b4250929ac6e6d2/pkg/util/log/file.go#L210)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> In `loggingT.outputLogEntry`, consider `!logDir.isSet()` to be equivalent to `l.toStderr`. Tests usually log to stderr by default anyone. The ones that don't should be special snowflakes which use `TestLogScope`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Indeed by having logging output go to stderr by default everything is much simplified. PTAL.

petermattis commented:
Review status: 4 of 10 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/cli/cli_test.go, line 80 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsdv1dEFXxjYPTN8v:-KWcsdv1dEFXxjYPTN8w:ba3dfna) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L80)):*
> ```Go
> 
> 	// Ensure that logging goes to a test-specific directory.
> 	//	c.logScope = log.MakeTestLogScope(2, c.fatalFn)
> ```

Remove?

---

*[pkg/cli/cli_test.go, line 90 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsz9-ojFAToBEbH2z:-KWcsz9-ojFAToBEbH3-:ba3dfna) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L90)):*
> ```Go
> 	s, err := serverutils.StartServerRaw(base.TestServerArgs{Insecure: insecure})
> 	if err != nil {
> 		//		c.logScope.Close(true, c.fatalFn)
> ```

Remove?

---

*[pkg/cli/cli_test.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsLa-yKRl8_pkI5ut:-KWcsLa-yKRl8_pkI5uu:b-kyfnzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L467)):*
> ```Go
> 	c.RunWithArgs([]string{"sql", "--alsologtostderr=false", "-e", "select 1"})
> 	c.RunWithArgs([]string{"sql", "--log-backtrace-at=foo.go:1", "-e", "select 1"})
> 	// c.RunWithArgs([]string{"sql", "--log-dir=", "-e", "select 1"})
> ```

What happens for this case now? Should it be removed?

---

*[pkg/util/log/clog_test.go, line 102 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcs4CzHf847IJKXFEN:-KWcs4CzHf847IJKXFEO:b-owwdw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/util/log/clog_test.go#L102)):*
> ```Go
> }
> 
> func logscope(t *testing.T) TestLogScope {
> ```

`s/logscope/logScope/g`

---

*[pkg/util/log/test_log_scope.go, line 46 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcslON3pVf73kYYmVF:-KWcslON3pVf73kYYmVG:b-u2a4rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/util/log/test_log_scope.go#L46)):*
> ```Go
> // lifetime of a logging directory. The logging directory is named
> // after the caller of MakeTestLogScope, up `skip` caller levels.
> func MakeTestLogScope(skip int, fatalFn func(string, ...interface{})) TestLogScope {
> ```

Does this need to be exported now? It doesn't seem to be used outside of this package unless I missed something. Can it live entirely in test code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 4 of 10 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/cli/cli_test.go, line 80 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsdv1dEFXxjYPTN8v:-KWcvCudBWMHw4I0hHzo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L80)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Remove?
</details>
Done.

---

*[pkg/cli/cli_test.go, line 90 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsz9-ojFAToBEbH2z:-KWcvDKtLjd-DwkAiaWt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L90)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Remove?
</details>
Done.

---

*[pkg/cli/cli_test.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsLa-yKRl8_pkI5ut:-KWcvDyRy_NSQWgip5Ur:b-du8vm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L467)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What happens for this case now? Should it be removed?
</details>
I don't know. I don't really understand what we want to happen here. 

---

*[pkg/util/log/clog_test.go, line 102 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcs4CzHf847IJKXFEN:-KWcwJ5GUOndrSYbC-2v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/util/log/clog_test.go#L102)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/logscope/logScope/g`
</details>
Done.

---

*[pkg/util/log/test_log_scope.go, line 46 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcslON3pVf73kYYmVF:-KWcwJmYUY9FbCVyJsnM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/util/log/test_log_scope.go#L46)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Does this need to be exported now? It doesn't seem to be used outside of this package unless I missed something. Can it live entirely in test code?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Hum although I un-exported and simplified the testLogScope now I get an import linter failure:
https://teamcity.cockroachdb.com/viewLog.html?buildId=48918&buildTypeId=Cockroach_UnitTests

why is that?

tamird commented:
Because you are linking the `testing` package into production code, apparently. This is because your file `test_log_scope` which is used only in tests, does not end in `_test` and thus is getting linked into non-test code.

---

Reviewed 2 of 8 files at r5, 2 of 3 files at r6.
Review status: 8 of 10 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsLa-yKRl8_pkI5ut:-KWczhJD5H4yK6S5h5yk:b-e8k00h) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L467)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I don't know. I don't really understand what we want to happen here. 
</details>
When was it added? My guess is that it wants to test logging to stderr in the presence of the log dir flag.

---

*[pkg/cli/cli_test.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcyOGZB-LVntu4Gv7M:-KWcyOG_2OaJ2N7peDTM:b-tzyagc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L61)):*
> ```Go
> 
> 	var errorFn func(...interface{})
> 	var fatalFn func(string, ...interface{})
> ```

Instead of all this, why don't you return an error? Alternatively:


```
errorFn := func (args ...interface{}) {
  log.Error(context.Background(), args...)
}
fatalfFn := func(f, string, args ...interface{}) {
  log.Fatalf(context.Background(), f, args...)
}
if t != nil {
  errorFn = t.Error
  fatalfFn = t.Fatalf
}
```

I don't know why you're using runtime.Goexit, but it doesn't seem right.

---

*[pkg/cli/cli_test.go, line 128 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWczUFgfxpKS2ti00qY:-KWczUFgfxpKS2ti00qZ:bxfcxsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L128)):*
> ```Go
> }
> 
> // stop cleans up after the test.
> ```

document the runStopper argument?

---

*[pkg/cli/cli_test.go, line 129 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWczbfADClmJVWw_-oQ:-KWczbfADClmJVWw_-oR:b381tbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L129)):*
> ```Go
> 
> // stop cleans up after the test.
> // The log files are only removed if the test has failed.
> ```

isn't this backwards?

---

*[pkg/util/log/test_log_scope.go, line 39 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-KmWKJGQCHHnm42P:-KWd-KmWKJGQCHHnm42Q:b-47fdj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L39)):*
> ```Go
> // after the caller of MaketestLogScope, up `skip` caller levels.
> func logScope(t *testing.T) testLogScope {
> 	_, _, f := caller.Lookup(1)
> ```

reduce the scope of `f`

---

*[pkg/util/log/test_log_scope.go, line 47 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-O2U1rJAW-Nb-2Nq:-KWd-O2U1rJAW-Nb-2Nr:bj0os8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L47)):*
> ```Go
> 	tempDir, err := ioutil.TempDir("", testName)
> 	if err != nil {
> 		t.Fatalf("cannot create temp dir: %s", err)
> ```

nit: `t.Fatal(err)`

---

*[pkg/util/log/test_log_scope.go, line 57 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-WHlntdOpKH3mhpa:-KWd-WHlntdOpKH3mhpb:b-g9b0ok) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L57)):*
> ```Go
> // close cleans up a testLogScope. The directory and its contents are
> // deleted, unless the test has failed and the directory is non-empty.
> func (l *testLogScope) close(t *testing.T) {
> ```

make this a value receiver if you're going to dereference it every time.

---

*[pkg/util/log/test_log_scope.go, line 70 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-RUiTWI19eNTAind:-KWd-RUiTWI19eNTAine:b4ipt00) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L70)):*
> ```Go
> 		// If the test failed, we keep the log files for further investigation,
> 		// but only if there were any.
> 		t.Errorf("test log files left over in: %s\n", *l)
> ```

no need for the newline.

---

*[pkg/util/log/test_log_scope.go, line 89 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-joEYVuTrjeLzHcH:-KWd-joEYVuTrjeLzHcI:b-7zjmpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L89)):*
> ```Go
> 	// the next logging event.
> 	logging.mu.Lock()
> 	defer logging.mu.Unlock()
> ```

this is making my inconsistent-lock-ordering spidey-sense tingle. Why can this be:

```
logDir.Lock()
logDir.name = dir
logDir.Unlock()

logging.mu.Lock()
err := logging.closeFilesLocked()
logging.mu.Unlock()
return err
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWd0F5GmoSw-nLLhYCH:bwxfxcw)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 8 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsLa-yKRl8_pkI5ut:-KWdAHoJPQzXVlvo5ewt:b-9q2oqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L467)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> When was it added? My guess is that it wants to test logging to stderr in the presence of the log dir flag.
</details>
So what happens for this case now?

---

*[pkg/util/log/test_log_scope.go, line 46 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcslON3pVf73kYYmVF:-KWdAU3Q46ONjbBrjuO8:b-mpb8a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/util/log/test_log_scope.go#L46)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
</details>
I think you need to rename this file to `log_scope_test.go`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for the hint. Fixed.

---

Review status: 8 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcsLa-yKRl8_pkI5ut:-KWdDxs6RUL066ujv8Z2:bldt1bx) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/cli/cli_test.go#L467)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> So what happens for this case now?
</details>
Before my changes, --log-dir="" did not do anything in this test case, because the test case does not trigger logging messages.
With the changes, --log-dir="" does not do anything either, because although it now redirects to stderr, there are still no logging messages.

I reverted this test back to its initial state, so we don't have to talk about it, but I think it's not testing much.

---

*[pkg/cli/cli_test.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcyOGZB-LVntu4Gv7M:-KWdD8LWQrCZqpTk1p5x:b-ol6km) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L61)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Instead of all this, why don't you return an error? Alternatively:
> 
> 
> ```
> errorFn := func (args ...interface{}) {
>   log.Error(context.Background(), args...)
> }
> fatalfFn := func(f, string, args ...interface{}) {
>   log.Fatalf(context.Background(), f, args...)
> }
> if t != nil {
>   errorFn = t.Error
>   fatalfFn = t.Fatalf
> }
> ```
> 
> I don't know why you're using runtime.Goexit, but it doesn't seem right.
</details>
It's to ensure the defers are run before the test terminates. That's what testing.Fatal does.

---

*[pkg/cli/cli_test.go, line 128 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWczUFgfxpKS2ti00qY:-KWdDHd0kfFghPV0OwUr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L128)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> document the runStopper argument?
</details>
Done.

---

*[pkg/cli/cli_test.go, line 129 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWczbfADClmJVWw_-oQ:-KWdDHxuBfK-sGzwNKIh:b1u8wpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L129)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> isn't this backwards?
</details>
Welp, yes. Thanks, fixed.

---

*[pkg/util/log/test_log_scope.go, line 46 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcslON3pVf73kYYmVF:-KWdDMI0ooavy36GNoyT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19f609144953dbee1fc908502212ccdb0b762165/pkg/util/log/test_log_scope.go#L46)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think you need to rename this file to `log_scope_test.go`.
</details>
Done.

---

*[pkg/util/log/test_log_scope.go, line 39 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-KmWKJGQCHHnm42P:-KWdDMdoDlr5WfHV_1mW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> reduce the scope of `f`
</details>
Done.

---

*[pkg/util/log/test_log_scope.go, line 47 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-O2U1rJAW-Nb-2Nq:-KWdDQEtk2VJ36qBF6HX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L47)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: `t.Fatal(err)`
</details>
Done.

---

*[pkg/util/log/test_log_scope.go, line 57 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-WHlntdOpKH3mhpa:-KWdDVKCN4RrH-yybsb1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L57)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> make this a value receiver if you're going to dereference it every time.
</details>
Done.

---

*[pkg/util/log/test_log_scope.go, line 70 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-RUiTWI19eNTAind:-KWdDfpCV8FgR0b-pRG_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> no need for the newline.
</details>
Done.

---

*[pkg/util/log/test_log_scope.go, line 89 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWd-joEYVuTrjeLzHcH:-KWdDlRLSZxRYY3Z5X_t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/util/log/test_log_scope.go#L89)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is making my inconsistent-lock-ordering spidey-sense tingle. Why can this be:
> 
> ```
> logDir.Lock()
> logDir.name = dir
> logDir.Unlock()
> 
> logging.mu.Lock()
> err := logging.closeFilesLocked()
> logging.mu.Unlock()
> return err
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWdD5KmcjI5WujeTns9:b-pzaj5h)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 7 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWdFwgJO-CiSiMFidt0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r7.
Review status: 8 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/cli_test.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcyOGZB-LVntu4Gv7M:-KWdR-XCWIiTzrJl-7nS:bwcv6vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L61)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> It's to ensure the defers are run before the test terminates. That's what testing.Fatal does.
</details>
OK, then we need a comment. Also, please consider the form above (not especially that `fatalFn` is renamed `fatalfFn`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 8 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cli/cli_test.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcyOGZB-LVntu4Gv7M:-KWe6urc33Jan0xjeqS6:b-a0pnmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L61)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK, then we need a comment. Also, please consider the form above (not especially that `fatalFn` is renamed `fatalfFn`).
</details>
Done.
I am not willing to use the `log` functions here because for all we know at this point a test may be redirecting logs to files and we wouldn't see any output.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cli/cli_test.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcyOGZB-LVntu4Gv7M:-KWe7Sili55nQAdD16Dd:b-m62xiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L61)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
> I am not willing to use the `log` functions here because for all we know at this point a test may be redirecting logs to files and we wouldn't see any output.
</details>
...why this extra function only called from one place? by the way, why can't you use `panic` instead of `runtime.Goexit`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/cli/cli_test.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcyOGZB-LVntu4Gv7M:-KWe81UEBQwSm3cagR8C:b7fkm4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L61)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ...why this extra function only called from one place? by the way, why can't you use `panic` instead of `runtime.Goexit`?
</details>
Separating the function makes it easier to understand what it does (via its comment).

Using panic is an interesting idea, but where would I write the recover? Is testing already doing recover?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675)*
<!-- Sent from Reviewable.io -->

tamird commented:
You don't need to recover.

On Tue, Nov 15, 2016 at 5:39 PM, kena <notifications@github.com> wrote:

> Review status: 9 of 10 files reviewed at latest revision, 1 unresolved
> discussion, some commit checks pending.
> ------------------------------
>
> *pkg/cli/cli_test.go, line 61 at r6
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWcyOGZB-LVntu4Gv7M:-KWe81UEBQwSm3cagR8C:b7fkm4q>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/a7e7f9d21301db49935cf2891c0c3df7cf1729f8/pkg/cli/cli_test.go#L61>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> ...why this extra function only called from one place? by the way, why
> can't you use panic instead of runtime.Goexit?
>
>
> Separating the function makes it easier to understand what it does (via
> its comment).
>
> Using panic is an interesting idea, but where would I write the recover?
> Is testing already doing recover?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/10675>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10675#issuecomment-260793246>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPM9yLeEB8N-O8htB21dJ9J5WcNHKks5q-jSPgaJpZM4KxV_w>
> .
>

knz commented:
But without a recover, a test that fails with Fatal() will also cause the stack trace to be printed. For now tests failing with Fatal just report a simple failure message. Is that what we want?

tamird commented:
Well, I think it's probably fine, don't you?

On Tue, Nov 15, 2016 at 5:42 PM, kena <notifications@github.com> wrote:

> But without a recover, a test that fails with Fatal() will also cause the
> stack trace to be printed. For now tests failing with Fatal just report a
> simple failure message. Is that what we want?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10675#issuecomment-260793958>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPEwtVI6iMev3A28D4ykOHyWOTCtUks5q-jVFgaJpZM4KxV_w>
> .
>

tamird commented:
By the way, this seems to support the earlier suggestion of just returning
an error from that function.

On Tue, Nov 15, 2016 at 5:42 PM, Tamir Duberstein <tamir@cockroachlabs.com>
wrote:

> Well, I think it's probably fine, don't you?
>
> On Tue, Nov 15, 2016 at 5:42 PM, kena <notifications@github.com> wrote:
>
>> But without a recover, a test that fails with Fatal() will also cause the
>> stack trace to be printed. For now tests failing with Fatal just report a
>> simple failure message. Is that what we want?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cockroachdb/cockroach/pull/10675#issuecomment-260793958>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABdsPEwtVI6iMev3A28D4ykOHyWOTCtUks5q-jVFgaJpZM4KxV_w>
>> .
>>
>
>

knz commented:
On 15-11-16 23:43, Tamir Duberstein wrote:
> By the way, this seems to support the earlier suggestion of just returning
> an error from that function.

Please, show me, supposing I modify `newCLITest` to return an error (I
know how to do this!), how do you propose to modify the Example
functions to do the right thing wrt errors?

Is this what you have in mind?

```
c, err := newCLITest(t)
if err != nil {
  panic(err)
}
```

tamird commented:
Yep!

On Tue, Nov 15, 2016 at 5:46 PM, kena <notifications@github.com> wrote:

> On 15-11-16 23:43, Tamir Duberstein wrote:
> > By the way, this seems to support the earlier suggestion of just
> returning
> > an error from that function.
>
> Please, show me, supposing I modify `newCLITest` to return an error (I
> know how to do this!), how do you propose to modify the Example
> functions to do the right thing wrt errors?
>
> Is this what you have in mind?
>
> ```
> c, err := newCLITest(t)
> if err != nil {
> panic(err)
> }
> ```
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10675#issuecomment-260794933>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPBdXQp-vinU2y48wfe3c2fBdDXRbks5q-jY6gaJpZM4KxV_w>
> .
>

knz commented:
Thanks for the suggestion! PTAL

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/cli_test.go, line 224 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-KWff56tCbXP1hrLn14W:-KWff56tCbXP1hrLn14X:b-nxs7fh) ([raw file](https://github.com/cockroachdb/cockroach/blob/601669ed7d5f93e8c36c99b4cc27e71c63838cce/pkg/cli/cli_test.go#L224)):*
> ```Go
> 		panic(err)
> 	}
> 	defer c.stop(false)
> ```

why doesn't this want to stop the stopper? (just needs a comment)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10675#-:-KWffOAxZqnEF3-luiRY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Done
